pkgs/unified_analytics/lib/src/session.dart

```diff
@@ -94,15 +69,25 @@ class Session {
 
   /// This will go to the session file within the dart-tool
   /// directory and fetch the latest data from the json to update
-  /// the class's variables
+  /// the class's variables. If the json file is malformed, a new
+  /// session file will be recreated
   ///
   /// This allows the session data in this class to always be up
   /// to date incase another tool is also calling this package and
   /// making updates to the session file
   void _refreshSessionData() {
-    final String sessionFileContents = _sessionFile.readAsStringSync();
-    final Map<String, Object?> sessionObj = jsonDecode(sessionFileContents);
-    _sessionId = sessionObj['session_id'] as int;
-    _lastPing = sessionObj['last_ping'] as int;
+    try {
+      final String sessionFileContents = _sessionFile.readAsStringSync();
```

Comment:
I'm not familiar with this code base, nor with `FileSystem`, but I would have expected that `readAsStringSync` would throw an exception if the file doesn't exist and that you'd need to handle that exception here as well.

Reply:
I tend to take a more cautious approach when it comes to files. I think the only safe (in the sense of not risking exceptions) approach to the file system is to assume that you can't know the state of the file system even if you've asked about it. The fact that you were told that a file existed at time T1 doesn't mean that it will still exist at time T2, even if those times are only separated by a fraction of a millisecond.

If you don't want this method to throw an exception, then you need to catch the exception and handle it. If you want callers of this method to handle the exception, then you need to document it.

Reply:
I see, so in the case that the file does not exist, it would throw a `PathNotFoundException` error since it can't find a file when it attempts to read it. For that case, I would want the same error handling as `FormatException` when the contents of the file are not able to be parsed as json.

Is best practice in dart to repeat a block like below?

```dart
    try {
      final String sessionFileContents = _sessionFile.readAsStringSync();
      final Map<String, Object?> sessionObj = jsonDecode(sessionFileContents);
      _sessionId = sessionObj['session_id'] as int;
      _lastPing = sessionObj['last_ping'] as int;
    } on FormatException {
      Initializer.createSessionFile(sessionFile: _sessionFile);

      final String sessionFileContents = _sessionFile.readAsStringSync();
      final Map<String, Object?> sessionObj = jsonDecode(sessionFileContents);
      _sessionId = sessionObj['session_id'] as int;
      _lastPing = sessionObj['last_ping'] as int;
    } on PathNotFoundException {
      Initializer.createSessionFile(sessionFile: _sessionFile);

      final String sessionFileContents = _sessionFile.readAsStringSync();
      final Map<String, Object?> sessionObj = jsonDecode(sessionFileContents);
      _sessionId = sessionObj['session_id'] as int;
      _lastPing = sessionObj['last_ping'] as int;
    }
```

Or is there a better approach? I can incorporate this into my next PR which is in draft right now

Reply:
Unless you want to handle a superclass of both classes, then you're required to have multiple `on` clauses. You _can_, however, put the code in the try blocks into a private (or local) method and invoke the same code from both `on` clauses.

Reply:
Would defining the method to parse the contents of the file and assign to the class variables be okay to have nested within the method? Example below

```dart
  void _refreshSessionData() {
    /// Using a nested function here to reduce verbosity
    void parseContents() {
      final String sessionFileContents = _sessionFile.readAsStringSync();
      final Map<String, Object?> sessionObj = jsonDecode(sessionFileContents);
      _sessionId = sessionObj['session_id'] as int;
      _lastPing = sessionObj['last_ping'] as int;
    }

    try {
      parseContents();
    } on FormatException {
      Initializer.createSessionFile(sessionFile: _sessionFile);

      parseContents();
    } on PathNotFoundException {
      Initializer.createSessionFile(sessionFile: _sessionFile);

      parseContents();
    }
  }
```

This way it shows the intention that it should only be used within the `_refreshSessionData()` private method.

Reply:
Yes, that's a reasonable solution. It's what I meant by a "local" method (though technically it's a local function).